test/context_flush_test.rb

```diff
@@ -0,0 +1,376 @@
+  end
+
+  # rubocop:disable Metrics/AbcSize
+  def test_partial_caterpillar
+    tracer = get_test_tracer
+    context_flush = get_context_flush
+    tracer.context_flush = context_flush
+
+    write1 = Minitest::Mock.new
+    expected = []
+    MIN_SPANS.times do |i|
+      expected << "a.#{i}"
+    end
+    (MAX_SPANS - MIN_SPANS).times do |i|
+      expected << "b.#{i}"
+    end
+    # We need to sort the values the same way the values will be output by the test transport
+    expected.sort!
+    expected.each do |e|
+      write1.expect(:call_with_name, nil, [e])
+    end
+
+    write2 = Minitest::Mock.new
+    expected = ['root']
+    MIN_SPANS.times do |i|
+      expected << "b.#{i + MAX_SPANS - MIN_SPANS}"
+    end
+    # We need to sort the values the same way the values will be output by the test transport
+    expected.sort!
+    expected.each do |e|
+      write2.expect(:call_with_name, nil, [e])
+    end
+
+    tracer.trace('root') do
+      MIN_SPANS.times do |i|
+        tracer.trace("a.#{i}") do
+        end
+      end
+      spans = tracer.writer.spans()
+      assert_equal(0, spans.length, 'nothing should be flushed, as max limit is not reached')
+      MAX_SPANS.times do |i|
+        tracer.trace("b.#{i}") do
+        end
+      end
+      spans = tracer.writer.spans()
+      # Let's explain the extra span here, what should happen is:
+      # - root span is started
+      # - then 99 spans (10 from 1st batch, 89 from second batch) are put in context
+      # - then the 101th comes (the 90th from the second batch) and triggers a flush of everything but root span
+      # - then the last 10 spans from second batch are thrown in, so that's 10 left + the root span
+      assert_equal(1 + MIN_SPANS, tracer.call_context.length, 'some spans should have been sent')
+      assert_equal(MAX_SPANS, spans.length)
+      spans.each do |span|
+        write1.call_with_name(span.name)
+      end
+      write1.verify
+    end
+
+    spans = tracer.writer.spans()
+    assert_equal(MIN_SPANS + 1, spans.length)
+    spans.each do |span|
+      write2.call_with_name(span.name)
+    end
+    write2.verify
+  end
+
+  # Test the tracer configure args which are forwarded to context flush only.
+  def test_tracer_configure
+    tracer = get_test_tracer
+
+    old_context_flush = tracer.context_flush
+    tracer.configure()
+    assert_equal(old_context_flush, tracer.context_flush, 'the same context_flush should be reused')
+
+    tracer.configure(min_spans_before_partial_flush: 3,
+                     max_spans_before_partial_flush: 3)
+
+    refute_equal(old_context_flush, tracer.context_flush, 'another context_flush should be have been created')
+  end
+
+  def test_tracer_hard_limit_overrides_soft_limit
+    tracer = get_test_tracer
+
+    context = tracer.call_context
+    tracer.configure(min_spans_before_partial_flush: context.max_length,
+                     max_spans_before_partial_flush: context.max_length,
```

Comment:
As of today, context max_length and context_flush boundaries are the same, but it's not necessarily that way. In the future, hard limit could be higher, and this test has no reason to fail test, it just tests that when hard and soft limit are the same, hard limit prevails.